app/src/main/java/org/wikipedia/diff/ArticleEditDetailsViewModel.kt

```diff
@@ -15,67 +14,128 @@ import org.wikipedia.dataclient.mwapi.MwQueryResponse
-        if (curTitle == null) {
-            curTitle = pageTitle
-            getWatchedStatus(pageTitle)
-            getRevisionDetails(pageTitle, revisionId)
-        }
+    init {
+        getWatchedStatus()
+        getRevisionDetails(revisionToId, revisionFromId)
     }
 
-    private fun getWatchedStatus(title: PageTitle) {
+    private fun getWatchedStatus() {
         viewModelScope.launch(CoroutineExceptionHandler { _, throwable ->
             watchedStatus.postValue(Resource.Error(throwable))
         }) {
             withContext(Dispatchers.IO) {
-                watchedStatus.postValue(Resource.Success(ServiceFactory.get(title.wikiSite).getWatchedStatus(title.prefixedText)))
+                watchedStatus.postValue(Resource.Success(ServiceFactory.get(pageTitle.wikiSite).getWatchedStatus(pageTitle.prefixedText)))
             }
         }
     }
 
-    fun getRevisionDetails(title: PageTitle, revisionId: Long) {
+    fun getRevisionDetails(revisionIdTo: Long, revisionIdFrom: Long = -1) {
         viewModelScope.launch(CoroutineExceptionHandler { _, throwable ->
             revisionDetails.postValue(Resource.Error(throwable))
         }) {
             withContext(Dispatchers.IO) {
-                val response = ServiceFactory.get(title.wikiSite).getRevisionDetailsAscending(title.prefixedText, 2, revisionId)
-                val revisions = response.query?.firstPage()!!.revisions
-                if (revisions.isNotEmpty()) {
-                    revisionDetails.postValue(Resource.Success(revisions))
+                if (revisionIdFrom >= 0) {
+                    var responseFrom: MwQueryResponse? = null
+                    var responseTo: MwQueryResponse? = null
+                    coroutineScope {
+                        launch {
+                            responseFrom = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsWithInfo(pageTitle.prefixedText, 2, revisionIdFrom)
+                        }
+                        launch {
+                            responseTo = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsWithInfo(pageTitle.prefixedText, 2, revisionIdTo)
+                        }
+                    }
+
+                    val pageTo = responseTo?.query?.firstPage()!!
+                    revisionFrom = responseFrom?.query?.firstPage()!!.revisions[0]
+                    revisionTo = pageTo.revisions[0]
+                    canGoForward = revisionTo!!.revId < pageTo.lastrevid
+                } else {
+                    val response = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsWithInfo(pageTitle.prefixedText, 2, revisionIdTo)
+                    val page = response.query?.firstPage()!!
+                    val revisions = page.revisions
+                    revisionTo = revisions[0]
+                    canGoForward = revisions[0].revId < page.lastrevid
+                    revisionFrom = if (revisions.size > 1) { revisions[1] } else null
```

Comment:
We can use `revisions.getOrNull(1)` here

Reply:
yep!